Mathlib/NumberTheory/SelbergSieve.lean

```diff
@@ -0,0 +1,237 @@
+  support : Finset ℕ
+  /-- The finite set of prime numbers whose multiples are to be sifted from `support`. -/
+  prodPrimes : ℕ
+  prodPrimes_squarefree : Squarefree prodPrimes
+  /-- A sequence representing how much each element of `support` should be weighted. -/
+  weights : ℕ → ℝ
+  weights_nonneg : ∀ n : ℕ, 0 ≤ weights n
+  /-- An approximation to `∑ i in support, weights i`, i.e. the size of the unsifted set. A bad
+    approximation will yield a weak statement in the final theorem. -/
+  totalMass : ℝ
+  /-- `nu d` is an approximation to the proportion of elements of `support` that are a multiple of
+    `d` -/
+  nu : ArithmeticFunction ℝ
+  nu_mult : nu.IsMultiplicative
+  nu_pos_of_prime : ∀ p : ℕ, p.Prime → p ∣ prodPrimes → 0 < nu p
+  nu_lt_one_of_prime : ∀ p : ℕ, p.Prime → p ∣ prodPrimes → nu p < 1
+  /-- The `level` of the sieve controls how many terms we include in the inclusion-exclusion type
+    sum. A higher level will yield a tighter bound for the main term, but will also increase the
+    size of the error term. -/
+  level : ℝ
+  one_le_level : 1 ≤ level
+
+attribute [arith_mult] SelbergSieve.nu_mult
+
+namespace SelbergSieve
+
+namespace Notation
+
+@[inherit_doc nu]
+scoped notation3 "ν" => nu
+@[inherit_doc prodPrimes]
+scoped notation3 "P" => prodPrimes
+@[inherit_doc weights]
+scoped notation3 "a" => weights
+@[inherit_doc totalMass]
+scoped notation3 "X" => totalMass
+@[inherit_doc support]
+scoped notation3 "A" => support
+@[inherit_doc level]
+scoped notation3 "y" => level
+/-- `1 ≤ y`-/
+scoped notation3 "hy" => one_le_level
+
+end Notation
+
+open Notation
+
+variable [s : SelbergSieve]
+
+/-! Lemmas aboud $P$. -/
+
+theorem prodPrimes_ne_zero : P ≠ 0 :=
+  Squarefree.ne_zero prodPrimes_squarefree
+
+theorem squarefree_of_dvd_prodPrimes {d : ℕ} (hd : d ∣ P) : Squarefree d :=
+  Squarefree.squarefree_of_dvd hd prodPrimes_squarefree
+
+theorem squarefree_of_mem_divisors_prodPrimes {d : ℕ} (hd : d ∈ divisors P) : Squarefree d := by
+  simp only [Nat.mem_divisors] at hd
+  exact Squarefree.squarefree_of_dvd hd.left prodPrimes_squarefree
+
+/-! Lemmas about $\nu$. -/
+
+theorem prod_primeFactors_nu {d : ℕ} (hd : d ∣ P) : ∏ p ∈ d.primeFactors, ν p = ν d := by
+  rw [← nu_mult.map_prod_of_subset_primeFactors _ _ subset_rfl,
+    Nat.prod_primeFactors_of_squarefree <| Squarefree.squarefree_of_dvd hd prodPrimes_squarefree]
+
+theorem nu_pos_of_dvd_prodPrimes {d : ℕ} (hd : d ∣ P) : 0 < ν d := by
+  calc
+    0 < ∏ p ∈ d.primeFactors, ν p := by
+      apply prod_pos
+      intro p hpd
+      have hp_prime : p.Prime := prime_of_mem_primeFactors hpd
+      have hp_dvd : p ∣ P := (dvd_of_mem_primeFactors hpd).trans hd
+      exact nu_pos_of_prime p hp_prime hp_dvd
+    _ = ν d := prod_primeFactors_nu hd
+
+theorem nu_ne_zero {d : ℕ} (hd : d ∣ P) : ν d ≠ 0 := by
+  apply _root_.ne_of_gt
+  exact nu_pos_of_dvd_prodPrimes hd
+
+theorem nu_lt_self_of_dvd_prodPrimes (d : ℕ) (hdP : d ∣ P) (hd_ne_one : d ≠ 1) : ν d < 1 := by
```

Comment:
is `nu_lt_self` the right name? I would think less than `self` means `ν d < d`; do you mean `nu_lt_one_`...?

Reply:
and can `d` be made implicit here, picked up in applications from `hdP`?

Reply:
Ah! `nu_lt_self` is a holdover from when I used the `nu d / d * X` convention.